cli/lib/webpack/service.js

```diff
@@ -0,0 +1,226 @@
+
+class WebpackService {
+	constructor(options) {
+		const { AppcdClient, builder, cli, logger } = options;
+		this.client = new AppcdClient();
+		this.logger = logger;
+		this.builder = builder;
+		const tiapp = cli.tiapp;
+		this.projectSettings = {
+			path: cli.argv['project-dir'],
+			name: tiapp.name,
+			type: tiapp.webpack.type,
+		};
+		this.loadBuildSettings(builder, cli);
+		this.shouldWatch = builder.deployType !== 'production' && !builder.buildOnly;
+
+		this.forceRebuild = this.buildSettings.platform === 'ios' ? builder.forceCleanBuild : builder.forceRebuild;
+		this.jobIdentifier = this.generateJobIdentifier();
+		this.pluginVersion = '1.0.0';
+		this.basePath = `/webpack/${this.pluginVersion}`;
+		this.webpackStatusEmitter = new EventEmitter();
+		this.webpackStatus = { state: 'unknown' };
+		this.reporter = new WebpackProgressReporter();
+	}
+
+	loadBuildSettings(builder, cli) {
+		const platform = cli.argv.platform;
+		const buildSettings = {
+			platform,
+			deployType: builder.deployType,
+			target: cli.argv.target,
+			sdk: {
+				path: cli.sdk.path,
+				version: cli.sdk.name,
+				gitHash: cli.sdk.manifest.githash,
+				buildDate: cli.sdk.manifest.timestamp
+			}
+		};
+
+		if (platform === 'ios') {
+			buildSettings.ios = {
+				deviceFamily: builder.deviceFamily
+			};
+		}
+
+		this.buildSettings = buildSettings;
+	}
+
+	ensureDaemonIsRunning() {
+		return new Promise((resolve, reject) => {
+			this.logger.debug('Connecting to appcd ...');
+			this.client.connect({ startDaemon: true })
+				.once('connected', () => {
+					this.logger.info(`🖥  Web UI ready on http://localhost:1732${this.basePath}/web`);
+					resolve();
+				})
+				.once('error', reject);
+		});
+	}
+
+	ensureWebpackIsRunning() {
+		const options = {
+			identifier: this.jobIdentifier,
+			watch: this.shouldWatch,
+			project: this.projectSettings,
+			build: this.buildSettings
+		};
+
+		return new Promise((resolve, reject) => {
+			if (this.forceRebuild) {
+				this.startWebpack(options).then(resolve).catch(reject);
+				return;
+			}
+
+			this.client
+				.request({
+					path: `${this.basePath}/status/${options.identifier}`,
+				})
+				.once('response', response => {
+					this.webpackStatus = response;
+					if (this.webpackStatus.state === 'stopped' || this.webpackStatus.state === 'error') {
+						this.logger.debug('Webpack not running, starting new Webpack build.');
+						this.startWebpack(options).then(resolve).catch(reject);
+					} else {
+						if (!this.validateOptions(response, options)) {
+							this.logger.debug('Build options changed, restarting Webpack build.');
+							this.startWebpack(options).then(resolve).catch(reject);
+							return;
+						}
+						if (this.webpackStatus.state === 'ready') {
+							this.builder.tiSymbols = response.tiSymbols;
+						}
+						return resolve();
+					}
+				})
+				.once('error', error => {
+					if (error.status === 404) {
+						this.startWebpack(options).then(resolve).catch(reject);
```

Comment:
<!--
  0 failure: 
  2 warning:  cli/lib/webpack/s..., cli/lib/webpack/s...
  
  
  DangerID: danger-id-Danger;
  File: cli/lib/webpack/service.js;
  Line: 103;
-->

- :warning: cli/lib/webpack/service.js line 103 – Avoid using promises inside of callbacks. (promise/no-promise-in-callback)
- :warning: cli/lib/webpack/service.js line 103 – Avoid using promises inside of callbacks. (promise/no-promise-in-callback)